nomad/core_sched.go

```diff
@@ -286,7 +288,20 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	gcEval := true
 	var gcAllocIDs []string
 	for _, alloc := range allocs {
-		if !alloc.TerminalStatus() || alloc.ModifyIndex > thresholdIndex {
+		if job == nil || job.Stop {
```

Comment:
@dadgar this part needs another look - here is where I addressed your comment about nil/stopped jobs, with a clarifying comment on why we should not consider a dead status job as GC eligibile.